src/transaction/log_page_buffer.c

```diff
@@ -2192,8 +2192,85 @@ logpb_verify_page_read (LOG_PAGEID pageid, const LOG_PAGE * left_log_pgptr, cons
+      for (;;)
+	{
+	  // if current log record is the last one, it must be the end of log
+	  if (left_log_rec_header->forw_lsa.is_null () || rite_log_rec_header->forw_lsa.is_null ())
+	    {
+	      // active transaction server and page server append the end of log with different transactions
+	      assert (left_log_rec_header->prev_tranlsa == rite_log_rec_header->prev_tranlsa
+		      && left_log_rec_header->back_lsa == rite_log_rec_header->back_lsa
+		      && left_log_rec_header->forw_lsa == rite_log_rec_header->forw_lsa
+		      //&& left_log_rec_header->trid == rite_log_rec_header->trid
+		      && left_log_rec_header->type == rite_log_rec_header->type);
+
+	      assert (left_log_rec_header->forw_lsa.is_null () && rite_log_rec_header->forw_lsa.is_null ());
+	      assert (left_log_rec_header->type == LOG_END_OF_LOG);
+
+	      // compare everything until end of log
+	      const char *const left_log_page_area = left_log_pgptr->area;
+	      const char *const rite_log_page_area = rite_log_pgptr->area;
+	      const int cmp_res = strncmp (left_log_page_area, rite_log_page_area, curr_left_lsa.offset
+					   + offsetof (LOG_RECORD_HEADER, trid));
+	      pages_equal = (cmp_res == 0);
+	      break;
+	    }
+
+	  // if not the last page, headers must be equal
+	  assert (*left_log_rec_header == *rite_log_rec_header);
+
+	  // if not the last log page, must stop, the pages are not equal
+	  if (left_log_rec_header->forw_lsa.pageid != curr_left_lsa.pageid
+	      || rite_log_rec_header->forw_lsa.pageid != curr_rite_lsa.pageid)
+	    {
+	      assert (left_log_rec_header->forw_lsa.pageid != curr_left_lsa.pageid
+		      && rite_log_rec_header->forw_lsa.pageid != curr_rite_lsa.pageid);
+
+	      // it is not this function's job to validate page corruption
+	      // if any of the pages is corrupted, withhold decision for log recovery analysis logic
+	      const bool left_has_valid_checksum = logpb_page_has_valid_checksum (left_log_pgptr);
+	      const bool rite_has_valid_checksum = logpb_page_has_valid_checksum (rite_log_pgptr);
+	      pages_equal = (!left_has_valid_checksum || !rite_has_valid_checksum);
+	      break;
+	    }
+
+	  // next log record still in the same page, advance to it
+	  curr_left_lsa = left_log_rec_header->forw_lsa;
+	  curr_rite_lsa = rite_log_rec_header->forw_lsa;
+
+	  left_log_rec_header = (const LOG_RECORD_HEADER *) (left_log_pgptr->area + curr_left_lsa.offset);
+	  rite_log_rec_header = (const LOG_RECORD_HEADER *) (rite_log_pgptr->area + curr_rite_lsa.offset);
+	}
+    }
+  else if (pageid == log_Gl.hdr.append_lsa.pageid)
+    {
+      // last page, not in recovery analysis
       const char *const left_log_page_area = left_log_pgptr->area;
       const char *const rite_log_page_area = rite_log_pgptr->area;
       const int cmp_res = strncmp (left_log_page_area, rite_log_page_area, log_Gl.hdr.append_lsa.offset);
```

Comment:
log_Gl.hdr.append_lsa may be higher on PTS than on ATS. Thus, latest log records may not arrived yet on PTS.